src/components/connectAdvanced.tsx

```diff
@@ -1,44 +1,63 @@
 import hoistStatics from 'hoist-non-react-statics'
-import React, { useContext, useMemo, useRef, useReducer } from 'react'
+import React, {
+  useContext,
+  useMemo,
+  useRef,
+  useReducer,
+  useLayoutEffect,
+} from 'react'
 import { isValidElementType, isContextConsumer } from 'react-is'
-import { createSubscription } from '../utils/Subscription'
+import type { Store } from 'redux'
+import type { SelectorFactory } from '../connect/selectorFactory'
+import { createSubscription, Subscription } from '../utils/Subscription'
 import { useIsomorphicLayoutEffect } from '../utils/useIsomorphicLayoutEffect'
 
-import { ReactReduxContext } from './Context'
+import {
+  ReactReduxContext,
+  ReactReduxContextValue,
+  ReactReduxContextInstance,
+} from './Context'
 
 // Define some constant arrays just to avoid re-creating these
-const EMPTY_ARRAY = []
+const EMPTY_ARRAY: [unknown, number] = [null, 0]
 const NO_SUBSCRIPTION_ARRAY = [null, null]
 
-const stringifyComponent = (Comp) => {
+const stringifyComponent = (Comp: unknown) => {
   try {
     return JSON.stringify(Comp)
   } catch (err) {
     return String(Comp)
   }
 }
 
-function storeStateUpdatesReducer(state, action) {
+function storeStateUpdatesReducer(
+  state: [payload: unknown, counter: number],
+  action: { payload: unknown }
+) {
   const [, updateCount] = state
   return [action.payload, updateCount + 1]
 }
 
+type UseLayoutEffect = typeof useLayoutEffect
+type EffectFunc = (...args: any[]) => void | ReturnType<React.EffectCallback>
+type EffectDeps = Parameters<UseLayoutEffect>[1]
```

Comment:
There is a `DependencyList` type exported from `react` if you'd like to use that instead!